[guilib] add template for GetWindow() to enclose type cast

xhaggi opened the pull request:
This PR will introduce a template for `CGUIWindowManager::GetWindow(int id)` to get rid of all the different type casts we currently use and to save some dupe code if using C++ style type casts.

It also adds a convenient method `CGUIWindowManager::GetDialog(int id)` which check if the window for the supplied ID is a dialog and directly return a `CGUIDialog*`. So no need to cast `CGUIWindow*` to `CGUIDialog*` in that cases, just use the new method.

I've done all the work to use the new generic `GetWindow<T>()` and `GetDialog()` where possible.

Runtime tested on MacOS

@Paxxi mind taking a look.

xhaggi committed:
[guilib] add template for GetWindow() to enclose type cast

xhaggi committed:
[refactor] use generic GetWindow() or GetDialog()

ksooo commented:
This is a nice one. Thanks.

xhaggi commented:
@Paxxi I've added a new commit (WIP) where I introduce a new map to receive the window id by type. With this new map and method it should be possible to get rid of the window id parameter. It would be nice if you could look at it and let me know if we're on the right direction.

xhaggi commented on xbmc/guilib/GUIWindowManager.h:
```diff
@@ -227,6 +231,7 @@ class CGUIWindowManager : public KODI::MESSAGING::IMessageTarget
 
   typedef std::map<int, CGUIWindow *> WindowMap;
   WindowMap m_mapWindows;
+  std::unordered_map<std::type_index, CGUIWindow*> m_mapWindowTypes;
```
I use an unordered_map here to optimize map lookups. We don't need an ordered map and only insert/remove items on start and shutdown of the app.

Four years ago a cache `CGUIWindowManagerIdCache` was introduced to optimize the map lookup of std::map in `GetWindow(int id)`. Not sure if he checked the performance of an unordered_map.

https://github.com/xbmc/xbmc/commit/98b4f9c083fc7c6ec1549228ac59e9d3b6ad9bbc

xhaggi committed:
[cleanup] various code style cleanups in GUIWindowManager

* use range based for loops where possible
* get rid of typedef's for iterators and use auto instead
* get rid of CGUIWindowManagerIdCache and use unordered_map instead
* use erase-remove idiom instead of for over iterators
* use nullptr instead of NULL

xhaggi commented:
I added another commit with various refactorings / cleanups applied to GUIWindowManager

xhaggi committed:
[guilib] add convenient method GetDialog() to GUIWindowManager

Paxxi commented on xbmc/addons/RepositoryUpdater.cpp:
```diff
@@ -113,7 +113,7 @@ bool CRepositoryUpdater::CheckForUpdates(bool showProgress)
 static void SetProgressIndicator(CRepositoryUpdateJob* job)
 {
-  auto* dialog = static_cast<CGUIDialogExtendedProgressBar*>(g_windowManager.GetWindow(WINDOW_DIALOG_EXT_PROGRESS));
+  auto* dialog = g_windowManager.GetWindow<CGUIDialogExtendedProgressBar>();
```
no need for `auto*` plain `auto` is enough but it already was there so

xhaggi replied:
not added by me ;) but will change it.

Paxxi commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -458,17 +458,17 @@ bool CGUIWindowManager::SendMessage(CGUIMessage& message)
     CSingleLock lock(g_graphicsContext);
-    for (rDialog it = m_activeDialogs.rbegin(); it != m_activeDialogs.rend(); ++it)
+
+    for (const auto& window : m_activeDialogs)
```
This changes behavior and may break in subtle ways, I'd leave it as is

xhaggi replied:
overlooked that a reverse iterator is used here.

xhaggi replied:
after a second look: I don't know why we force reverse order to call OnMessage() .. it doesn't matter if you ask me. 2  lines below we iterate over all windows without a strict order ;)

Paxxi commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -561,18 +561,19 @@ void CGUIWindowManager::Add(CGUIWindow* pWindow)
+  for (int id : pWindow->GetIDRange())
   {
-    WindowMap::iterator it = m_mapWindows.find(*idIt);
+    const auto& it = m_mapWindows.find(id);
```
no need for `const auto&` plain `auto` is all you need for the iterator returned

Paxxi commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -561,18 +561,19 @@ void CGUIWindowManager::Add(CGUIWindow* pWindow)
     }
-    m_mapWindows.insert(std::pair<int, CGUIWindow *>(*idIt, pWindow));
+
+    m_mapWindows.insert(std::pair<int, CGUIWindow*>(id, pWindow));
```
use `std::make_pair(id, pWindow)` instead of `std::pair<int, CGUIWindow*>(id, pWindow)`
and same goes for next line

xhaggi replied:
c/p ;) but you're right

Paxxi commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -598,18 +599,16 @@ void CGUIWindowManager::RegisterDialog(CGUIWindow* dialog)
-  m_idCache.Invalidate();
-  WindowMap::iterator it = m_mapWindows.find(id);
+
+  const auto& it = m_mapWindows.find(id);
```
plain `auto` is enough

Paxxi commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -1174,32 +1173,45 @@ void CGUIWindowManager::FrameMove()
-    return window;
-
-  WindowMap::const_iterator it = m_mapWindows.find(id);
+  const auto& it = m_mapWindows.find(id);
```
plain `auto` is enough

Paxxi reviewed (commented):
I would go full auto wherever GetWindow or GetDialog is used to avoid having to repeat the type twice but not important enough for hundreds of changes that I require it 😄 

I don't think the use of const auto& in most for loops are actually of any use since the vectors contain pointers a plain auto should be perfectly fine to use.

Also for someone claiming not to be a c++ dev you sure know some stuff, I did not know about std::type_index 😄 

Overall really nice work, 👍

xhaggi commented:
@Paxxi I've addressed your comments and squashed some commits.

xhaggi commented:
jenkins build this please

Montellese commented on xbmc/guilib/GUIWindowManager.h:
```diff
@@ -31,6 +31,8 @@
 #include <list>
 #include <utility>
 #include <vector>
+#include <typeindex>
```
Can you add the includes in alphabetical order?

xhaggi replied:
Sure

xhaggi commented:
@Montellese the includes are now in alphabetic order.

Paxxi commented:
We had a brief discussion about this on slack and it's all good with this approach

xhaggi commented:
Small update: I've dropped the is_pointer check from the template and instead I've added a check if the supplied type derives from `CGUIWindow`. `std::is_base_of` will guard against the pointer issue too. I updated the code doc for the template too.

xhaggi commented:
jenkins build and merge

Montellese commented:
I actually wanted to propose `std::is_base_of` but I figured it would get too complicated but great that you use it. The only thing I don't like about these type traits is that they produce error messages that nobody understands. That's why I usually combine them with a `static_assert()` like
```
template<typename T, typename std::enable_if<std::is_base_of<CGUIWindow,T>::value>::type* = nullptr>
T* GetWindow() const
{
  static_assert(std::is_base_of<CGUIWindow,T>::value, "Type not derived from CGUIWindow in CGUIWindowManager::GetWindow()");
  return dynamic_cast<T *>(GetWindow(std::type_index(typeid(T))));
};
```
That way the compiler will create an error with the message "Type not derived from CGUIWindow in CGUIWindowManager::GetWindow()" instead of a mumbo jumbo of template parameters and instantiations across multiple lines.

xhaggi commented:
@Montellese meh too late .. but I will do another PR with your suggestion.

popcornmix commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -1196,6 +1196,14 @@ CGUIWindow* CGUIWindowManager::GetWindow(std::type_index type) const
+CGUIDialog* CGUIWindowManager::GetDialog(int id) const
+{
+  CGUIWindow *window = GetWindow(id);
+  if (window->IsDialog())
```
I'm hitting a segfault here with window=null. Changing to 
```if (window && window->IsDialog())```
avoids the crash. See: http://forum.kodi.tv/showthread.php?tid=310869

xhaggi replied:
You're right missed that.

ksooo commented:
@xhaggi this unfortunately only works if there is at most one instance of the given window type, which is not always the case for some pvr windows. => https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIWindowManager.cpp#L260 For those windows, this PR leads to some serious regressions, because now a "random" window gets returned instead of the requested tv or radio instance of the window type.

As I never liked the strange reusage of the same window type for tv and radio windows (this violates OOP principles imo) I will fix this for the respective PVR windows.

But I want to give a heads-up here that the new template function must only be used if one can be sure that at a given point in time there is at least one instance of the requested type. If there are other reused window types, this PR might have also introduced regressions.

xhaggi commented:
Thanks for pointing it out. Never thought of that while @Paxxi and me talked about a way to get rid of the window id in the template. The first approach was a template where you need to pass in the window id and the generic return type.

```
template<typename T, typename std::enable_if<std::is_base_of<CGUIWindow,T>::value>::type* = nullptr>
T* GetWindow(int id) const { return dynamic_cast<T *>(GetWindow(id)); };
```

IMO we should introduce this template too and use it where we can't use the other, or don't we?

xhaggi commented:
Or we drop the current template and pass in the id everywhere to be safe.

ksooo commented:
I don't think it is a good idea to forbid creation of multiple instances of a window class. This introduces less flexibility just to make this single template function work. But if we allow multiple instances, the template you introduced here is really dangerous to use. 

> Or we drop the current template and pass in the id everywhere to be safe.

 Yes, this is what needs to be done.

ksooo commented:
Unfortunately, this is some real monkey work, going to every change you made in this PR, reintroducing the id you removed before. :-/

xhaggi commented:
Will create a PR for it tomorrow if I found some spare time.

ksooo commented:
> Will create a PR for it tomorrow if I found some spare time.

 Thanks. Much appreciated.
